pkg/prometheus/agent/operator.go

```diff
@@ -0,0 +1,1563 @@
+	errs := make([]error, 0)
+
+	for _, n := range nodes.Items {
+		address, _, err := nodeAddress(n)
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "failed to determine hostname for node (%s)", n.Name))
+			continue
+		}
+		addresses = append(addresses, v1.EndpointAddress{
+			IP: address,
+			TargetRef: &v1.ObjectReference{
+				Kind:       "Node",
+				Name:       n.Name,
+				UID:        n.UID,
+				APIVersion: n.APIVersion,
+			},
+		})
+	}
+
+	return addresses, errs
+}
+
+// nodeAddresses returns the provided node's address, based on the priority:
+// 1. NodeInternalIP
+// 2. NodeExternalIP
+//
+// Copied from github.com/prometheus/prometheus/discovery/kubernetes/node.go
+func nodeAddress(node v1.Node) (string, map[v1.NodeAddressType][]string, error) {
+	m := map[v1.NodeAddressType][]string{}
+	for _, a := range node.Status.Addresses {
+		m[a.Type] = append(m[a.Type], a.Address)
+	}
+
+	if addresses, ok := m[v1.NodeInternalIP]; ok {
+		return addresses[0], m, nil
+	}
+	if addresses, ok := m[v1.NodeExternalIP]; ok {
+		return addresses[0], m, nil
+	}
+	return "", m, fmt.Errorf("host address unknown")
+}
+
+// Resolve implements the operator.Syncer interface.
+func (c *Operator) Resolve(ss *appsv1.StatefulSet) metav1.Object {
+	key, ok := c.accessor.MetaNamespaceKey(ss)
+	if !ok {
+		return nil
+	}
+
+	match, promKey := prompkg.StatefulSetKeyToPrometheusKey(key)
+	if !match {
+		level.Debug(c.logger).Log("msg", "StatefulSet key did not match a Prometheus key format", "key", key)
+		return nil
+	}
+
+	p, err := c.promInfs.Get(promKey)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+
+	if err != nil {
+		level.Error(c.logger).Log("msg", "Prometheus lookup failed", "err", err)
+		return nil
+	}
+
+	return p.(*monitoringv1.PrometheusAgent)
+}
+
+// Sync implements the operator.Syncer interface.
+func (c *Operator) Sync(ctx context.Context, key string) error {
+	err := c.sync(ctx, key)
+	c.reconciliations.SetStatus(key, err)
+
+	return err
+}
+
+func (c *Operator) sync(ctx context.Context, key string) error {
+	pobj, err := c.promInfs.Get(key)
+
+	if apierrors.IsNotFound(err) {
+		c.reconciliations.ForgetObject(key)
+		// Dependent resources are cleaned up by K8s via OwnerReferences
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	p := pobj.(*monitoringv1.PrometheusAgent)
+	p = p.DeepCopy()
+	if err := k8sutil.AddTypeInformationToObject(p); err != nil {
+		return errors.Wrap(err, "failed to set Prometheus type information")
+	}
+
+	logger := log.With(c.logger, "key", key)
+	if p.Spec.Paused {
+		level.Info(logger).Log("msg", "the resource is paused, not reconciling")
+		return nil
+	}
+
+	level.Info(logger).Log("msg", "sync prometheus")
+
+	cg, err := prompkg.NewConfigGenerator(c.logger, p, c.endpointSliceSupported)
+	if err != nil {
+		return err
+	}
+
+	assetStore := assets.NewStore(c.kclient.CoreV1(), c.kclient.CoreV1())
+	if err := c.createOrUpdateConfigurationSecret(ctx, p, cg, assetStore); err != nil {
```

Comment:
Isn't there a risk of name clashing if both "foo" Prometheus and "foo" PrometheusAgent exist in the same namespace? The same would be true for all dependent resources managed by the controller.

Reply:
Oh, thats a great catch! Let's name agent pods as prom-agents(or similar) to avoid that :)